### PR TITLE
Add `--show-context` to kw explore

### PR DESCRIPTION
**Pull request by sahil-sagwekar2652**
The issue proposed to use the batcat command but this commit uses the built-in -C flags in git grep and GNU grep instead, to avoid adding more dependencies to kw.

This commit also updates the docs as well as the tests for kw-explore.

Is the choice of using `-C` appropriate? Or would it be confusing because there is also the `-c` which exists. Please suggest your thoughts on this.

Closes: #1052

### CODE REVIEW

**Review by davidbtadokoro — CHANGES_REQUESTED** [review]
Hi @sahil-sagwekar2652, and thank you so much for your first change!

As starting feedback, it seems you get the gist of the contribution practices and processes (which aren't trivial), so great job!

I left some review comments, and I recommend you visualize them by going into the `Commits` tab and seeing my comments with context (kinda meta considering the purpose of your PR :rofl: ). Most are about coding style but are as important as any other, so see if you can tackle them. Feel free to ping me if you have doubts.

Also, here is a comment about the commit message that doesn't fit the inline review. Think that the issue you are alluding to won't be accessible when the change gets merged into the project, so make your commit message auto-contained, i.e., describe what, how, and why you are doing it (see [here](https://kworkflow.org/content/howtocontribute.html#commit-messages) for reference).

Apart from all of this, I tested the change in the three modalities (git grep, git grep with all files, and GNU grep), and it works as intended! @rodrigosiqueira, when you have the opportunity, check to see if this PR introduced the behavior you described in the issue.

File: src/explore.sh
Comment (sahil-sagwekar2652) [reply src/explore.sh:52]: I did think of this, but then I thought that `flag` being number 3 would maintain some consistency in all the functions, including `explore_git_log`.

I agree with your opinion too. Do let me know your final thoughts, I'm happy to go either way.

File: src/explore.sh
Comment (sahil-sagwekar2652) [reply src/explore.sh:234]: Yes, I actually wanted to confirm this. I understand now.
Similarly, if I understand correctly, I don't think we are following [this rule](https://kworkflow.org/content/codingstyle.html#avoid-short-options) of the coding style which suggests using long options instead of short ones, wherever available. 

According to it the command should look like - `git grep --context ${context} -e '${regex}' --line-number -I ${path}`

Comment (sahil-sagwekar2652) [reply src/explore.sh:234]: Yes, I actually wanted to confirm this. I understand now.
Similarly, if I understand correctly, I don't think we are following [this rule](https://kworkflow.org/content/codingstyle.html#avoid-short-options) of the coding style which suggests using long options instead of short ones, wherever available. 

According to it the command should look like - `git grep --context ${context} -e '${regex}' --line-number -I ${path}`

Comment (davidbtadokoro) [reply src/explore.sh:234]: Nice catch! You are correct, this rule should also be enforced, so take the opportunity to do this here (and in the other two functions).

Also, just as a reminder, you probably will find many other places in `src/explore.sh` that don't follow the coding style, but **don't change them** in this PR, as they would be outside the scope of the change.

**Comment by davidbtadokoro** [comment]
Hi @sahil-sagwekar2652, I see that you've requested a review, and I imagine you did it because of the comment I just answered. Keep in mind that we maintainers usually take a couple of days or more to look into new PRs or updates.

I'll seize the opportunity to add another comment I remembered I need to add to the review. As you added a new option to the `explore` feature, you must add Bash and Zsh completions. For reference, see [here](https://kworkflow.org/content/project_structure.html#adding-a-new-feature), but you can just analyze the `src/bash_autocomplete.sh` and `src/_kw` files to understand how the completions work (don't forget to test **both** completions, i.e., if you only use Bash or Zsh, install the other shell, install this PR version of kw and validate that the completion is the correct one).

**Comment by sahil-sagwekar2652** [comment]
Hello @davidbtadokoro, I mistakenly asked for the review, forgetting that this is an ongoing review. I apologise for any inconvenience. I have modified the code according to your suggestions.

I want to take this opportunity to reach out to you for the special ID required to submit a GSoC application.
Apologies for this digression I've been trying to contact you by email for the past couple of weeks.

Thanks in advance!

**Comment by davidbtadokoro** [comment]
> Hello @davidbtadokoro, I mistakenly asked for the review, forgetting that this is an ongoing review. I apologise for any inconvenience. I have modified the code according to your suggestions.

No worries! Sorry if I sounded annoyed, that really wasn't the case :smile:! I will review this new version ASAP. 

> 
> I want to take this opportunity to reach out to you for the special ID required to submit a GSoC application. Apologies for this digression I've been trying to contact you by email for the past couple of weeks.

Oops! I'm sorry if your message slipped through me. Funny enough, I can't find these messages :sweat:. I will send you an email with some information. 

> 
> Thanks in advance!

**Comment by davidbtadokoro** [comment]
Hi @sahil-sagwekar2652, and thank you once again for your update. From what I saw, you tackled all my comments.

However, you've updated the PR by adding commits on top of the original one, so you should squash all three commits into one. In case this doesn't make sense, think that there are two important premisses when dealing with PRs:

1. There is no problem in rewriting the commit history as a PR isn't merged code into the upstream yet (this **is** a great problem if we are talking about pushed code that is being used);
2. Every commit should be atomic in its sense. Each commit's "actual" code should be accompanied by related tests and documentation. Also, each commit should not fail in the CI jobs.

In your update, the second commit fixes the "problems" of the first by addressing my comments, and the third commit fixes the problems of the second (number 2 fails in the unit tests). So, as we have a chance to "rewrite" the first commit, we can refine the change into a single self-contained commit. Overall, when updating a PR, you should use tools like `git rebase --interactive` and `git commit --amend`, and if you are unfamiliar with these concepts, take a look at this [post](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) in the Git doc.

In any case, this PR is really promising, and besides my verbose comments, you are doing great!!

**Comment by sahil-sagwekar2652** [comment]
Hi @davidbtadokoro, thanks for the review. I have squashed the commits into a single commit.
Thanks again for all the help!

File: tests/unit/explore_test.sh
Comment (rodrigosiqueira) [inline tests/unit/explore_test.sh:75]: This type of test is okay, but I would prefer something that uses the feature. For example, it would be better to search for something in one of the sample files, capture the output, and check if it worked as expected.

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]
Hi @sahil-sagwekar2652 ,
Very nice PR! You can see some of my comments in the PR.

File: src/explore.sh
Comment (sahil-sagwekar2652) [reply src/explore.sh:168]: There are other instances like line `150` where double quotes are used. Should I change them too?

File: documentation/man/features/kw-explore.rst
Comment (sahil-sagwekar2652) [reply documentation/man/features/kw-explore.rst:49]: Hi @rodrigosiqueira, 

Thank you for the review and suggestions.
I do like the idea of defining a default value, but it brings up some inconsistencies.
From what I understand (I might be wrong), the `::` notation to define optional arguments doesn't work for short options. Additionally, I read on StackOverflow that it works in `getopts.c` but not in the bash version of getopts.

Using the `::` notation for `C` and `show-context` and changing the code inside the case, we get the following viable combinations:

1. `kw e -C string` and `kw e --show-context string` - uses the default context value, i.e., 3.
2. `kw e --show-context=5 string` - shows 5 lines of context.
3. `kw e -C=3 string` will not work.

**Comment by sahil-sagwekar2652** [comment]
Hello @rodrigosiqueira, I will squash this into one commit after the final approval.

File: documentation/man/features/kw-explore.rst
Comment (rodrigosiqueira) [reply documentation/man/features/kw-explore.rst]: I'm ok with `-C3`.

File: src/explore.sh
Comment (sahil-sagwekar2652) [reply src/explore.sh]: Hi @rodrigosiqueira , Double quotes are also being used throughout the switch case block to set the `options_values['ERROR']`
I think it would be a good idea to use single quotes in other cases too.

File: documentation/man/features/kw-explore.rst
Comment (rodrigosiqueira) [inline documentation/man/features/kw-explore.rst:50]: How about creating an example section and adding one or two examples of this feature?

File: src/_kw
Comment (rodrigosiqueira) [inline src/_kw:409]: Is this change part of this PR, or is this unrelated?

File: src/explore.sh
Comment (rodrigosiqueira) [inline src/explore.sh:177]: Don't you need to use `shift 2` since it can have a second item?

File: tests/unit/explore_test.sh
Comment (rodrigosiqueira) [inline tests/unit/explore_test.sh:79]: Use `--lines` instead of `-l`.

File: tests/unit/explore_test.sh
Comment (rodrigosiqueira) [inline tests/unit/explore_test.sh:80]: Counting the number of lines is good, but you could also add some tests that check for the string before and after the match. You can leverage some of the functions in the `test/unit/utils.sh`; for example, `assert_substring_match` could be handy for this type of test, but we have other options.

File: tests/unit/explore_test.sh
Comment (rodrigosiqueira) [inline tests/unit/explore_test.sh:73]: btw, the explore test has some bad test patterns, such as adding many checks in the same function. Since you are adding a new feature, I recommend you create specific functions for your tests. For example, you could create a function named `test_explore_with_git_and_context`. Ideally, we want one test function per use case since this keeps things easy to debug and update.

File: tests/unit/explore_test.sh
Comment (rodrigosiqueira) [inline tests/unit/explore_test.sh:264]: Use `(${LINENO})` in all of the assertions, and you don't need quotes around numbers.

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]

File: src/_kw
Comment (sahil-sagwekar2652) [reply src/_kw:409]: No, this shouldn't be here.

**Comment by sahil-sagwekar2652** [comment]
Hello, @rodrigosiqueira 
I'll send a commit ASAP, was caught up in college work.

File: src/explore.sh
Comment (sahil-sagwekar2652) [reply src/explore.sh:177]: There is a `shift` statement outside the else block. It adds up to have the same effect.

File: tests/unit/explore_test.sh
Comment (sahil-sagwekar2652) [reply tests/unit/explore_test.sh:297]: I think this change would be out of place for this PR. Would it be a good idea to change this in another commit?

**Comment by sahil-sagwekar2652** [comment]
Hello @rodrigosiqueira ,
I have added a new test function and also added some examples in the documentation.
I have not tackled [this](https://github.com/kworkflow/kworkflow/pull/1058#discussion_r1550670863) comment of yours in this commit.

**Comment by sahil-sagwekar2652** [comment]
Hello @rodrigosiqueira ,
I have added a new test function and also added some examples in the documentation.
I have not tackled [this](https://github.com/kworkflow/kworkflow/pull/1058#discussion_r1550670863) comment of yours in this commit.

File: tests/unit/explore_test.sh
Comment (davidbtadokoro) [reply tests/unit/explore_test.sh:297]: I think Siqueira was alluding to changing all occurrences in the additions of this PR as we are concatenating the parenthesis with the value of `LINENO`. The pattern in this file is misleading, so don't use it as basis :smile:

File: src/_kw
Comment (davidbtadokoro) [inline src/_kw:285]: The short option accepts arguments that come directly after, like in `-C5`, and the long option accepts arguments after the equal sign, like in `--show-context=5`. However, the Zsh completions leave a blank space after completing `-C` and `--show-context,` both of which will result in a parser error and a bad experience for the user. To fix this, change the line to

```bash
'(-C --show-context)'{-C-,--show-context=-}'[set the context value]' \
```

with the minus after `-C` and an equal sign and minus after `--show-context` inside the curly braces.

You probably shouldn't know this if you never played with Zsh completions (which aren't intuitive), so don't worry.

**Review by davidbtadokoro — CHANGES_REQUESTED** [review]
First, thanks again for the update and dedication to this change (it looks neat).

I've reviewed the comments of @rodrigosiqueira, and it seems that you've tackled all of them besides one, which I commented on. I've resolved the conversations that were clearly resolved but left the ones that can be interpretative open for Siqueira to check if they follow. I also left a comment which is more about how the native completions system for Zsh is wild and obscure :sweat_smile:

**Comment by sahil-sagwekar2652** [comment]
Hello @davidbtadokoro ,
I have changed `($LINENO)` to `(${LINENO})` only in context of my changes. We could change the other occurences in a separate PR. And I've also modified the `_kw` file according to your suggestion, thanks again.

**Comment by davidbtadokoro** [comment]
Hi @sahil-sagwekar2652, and thank you so much for the update! @rodrigosiqueira, I know I've already said this in this PR, but now I think everything looks legit on my end, so will wait for your green light. Tested it again in default, grep and all mode, and both options (short and long) work great (and the Zsh completions as well).

File: tests/unit/explore_test.sh
Comment (rodrigosiqueira) [inline tests/unit/explore_test.sh:193]: Why are you using upper case here? Lowercase makes more sense here.

File: tests/unit/explore_test.sh
Comment (rodrigosiqueira) [inline tests/unit/explore_test.sh:207]: Use single quotes in the string and add double quotes around `${LINENO}`.

File: tests/unit/explore_test.sh
Comment (rodrigosiqueira) [inline tests/unit/explore_test.sh:212]: use `printf` instead of `echo`.
https://kworkflow.org/content/codingstyle.html#use-printf-instead-of-echo

Also use --lines= instead of -n

and add double quotes around CONTEXT_LINE.

File: tests/unit/explore_test.sh
Comment (rodrigosiqueira) [inline tests/unit/explore_test.sh:213]: add `{}` around `CONTEXT_LINE`.

File: tests/unit/explore_test.sh
Comment (rodrigosiqueira) [inline tests/unit/explore_test.sh:221]: Check my previous comments since they also apply here.

File: tests/unit/explore_test.sh
Comment (rodrigosiqueira) [inline tests/unit/explore_test.sh:302]: Could you also add a test with an invalid option?

File: src/explore.sh
Comment (rodrigosiqueira) [inline src/explore.sh:292]: Since num is optional, add [] around it (e.g., --show-context[=<num>]

File: documentation/man/features/kw-explore.rst
Comment (rodrigosiqueira) [inline documentation/man/features/kw-explore.rst:48]: Since num is optional, add [] around it, e.g., --show-context[=<num>].

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]

**Comment by sahil-sagwekar2652** [comment]
Hello @rodrigosiqueira , I have tried to address all your comments, except the one for using `--lines` instead of `-n`,
I thought using `-n` with the `head` and `tail` commands is a well-known convention which is why I went with the short-hand flag.

There are many more instances of using the short flag with `head` and `tail` within the file, if we are going to be updating it to `--lines` should we do it for all occurrences? Please let me know your thoughts on this. Thanks.

**Comment by rodrigosiqueira** [comment]
Hi @sahil-sagwekar2652 ,
First of all, I'm sorry for my huge delay in reviewing this change. Anyway, it looks great! I already merged it into the unstable branch.

Thanks a lot.

**Pull request closed without merging**
